File: Hara-Chap05-regression.py
# パッケージをインポート
import pandas as pd
import statsmodels.api as sm


# データを読み込む
input_data = pd.read_csv("carprice.csv")

# 単回帰分析を行う。最小二乗法で推定する 
model = sm.OLS(input_data.price, sm.add_constant(input_data.enginesize))
results = model.fit()
print(results.summary())
print('p-values\n', results.pvalues)


#モデルで利用する変数を接合し, ひとつの DataFrame にする
#読み込むデータを変えれば, 説明変数や被説明変数を変えることが可能
x_list = pd.concat([input_data.price, input_data.enginesize,input_data.wheelbase, input_data.horsepower, input_data.carheight], axis=1)

x_list2 = x_list.drop("price", 1)
price = input_data.price

#重回帰分析の結果を表示する
model = sm.OLS(price, sm.add_constant(x_list2))
result =model.fit()
print(result.summary())
print(result.pvalues)

#ダミー変数を get_dummies で作成する
input_data2=pd.get_dummies(input_data)
#変数リストをプロットする
print(input_data2.columns.values)


# ダミー変数を組み込んだ回帰分析
from sklearn import linear_model, datasets
from sklearn.linear_model import LinearRegression

#モデルで利用する変数を接合し, ひとつの DataFrame にする
#読み込むデータを変えれば, 説明変数や被説明変数を変えることが可能
x_list3 = pd.concat([input_data2.enginesize,input_data2.wheelbase, input_data2.horsepower, input_data2.carheight, input_data2.doornumber_four, input_data2.fueltype_gas], axis=1)
price2 = input_data2.price


#重回帰分析の結果を表示する
model = sm.OLS(price2, sm.add_constant(x_list3))
result =model.fit()
print(result.summary())

print(result.pvalues)

#多重共線性が発生していないか確認する
from statsmodels.stats.outliers_influence import * num_cols = model.exog.shape[1] 
print(num_cols) #説明変数の列数
# VIF
vifs = [variance_inflation_factor(model.exog, i) for i in range(0, num_cols)]
pdv = pd.DataFrame(vifs, index=model.exog_names, columns=["VIF"])
print(pdv)

File: Hara-Chap05-regression.r
library(tidyverse) 
library(broom) # 分析結果を見やすくする
library(car) # vifの計算のため

# options(scipen=10) #指数表記を回避

# データを読み込む
setwd("/Users/soichi/Dropbox/Programing/Hara-Python/Hara-Python-1")
input_data <- readr::read_csv("carprice.csv")
# 先頭10行を確認
head(input_data,10)

# 関数lm()で線形回帰モデルをOLS推定する
model01 <- price ~ enginesize # 単回帰モデル
res01 <- lm(model01, data =input_data) # 最小二乗法
# 結果を表示
summary(res01)
tidy(res01)
tidy(res01)$p.value

# 重回帰
x_list <- input_data %>% dplyr::select(price, enginesize, wheelbase, horsepower, carheight)
res02 <- lm(price ~. , data = x_list)
tidy(res02)


# ダミー変数
## Rはファクター方の変数を自動でダミーにしてくれる
input_data$doornumber <- as.factor(input_data$doornumber)
x_list2 <- input_data %>% dplyr::select(price, enginesize, wheelbase, horsepower, carheight,doornumber)
res2 <- lm(price ~., data = x_list2)
tidy(res2)

# VIF
vif(res2) # vifを表示

# おまけ
## tidy形式で複数のモデルを効率的に管理して，あとで結果を活用しやすくする。
models <- c(
  price ~ enginesize,
  price ~ enginesize + wheelbase + horsepower + carheight,
  price ~ enginesize + wheelbase + horsepower + carheight + doornumber
) %>% 
  enframe("model_no", "models")

df_res <- models %>% 
  mutate(model = map(models, lm, data=input_data),
         tidied = map(model, tidy),
         glanced = map(model, glance)
         )

df_coef <- df_res %>% 
  select(model_no, tidied) %>% 
  unnest(cols=c(tidied)) %>% 
  mutate_if(is.double, round, digits=2)

df_coef %>% 
  mutate(term = fct_inorder(term)) %>% 
  select(model_no, term, estimate) %>% 
  spread(model_no,estimate)
